src/csharp/Microsoft.Spark/Broadcast.cs

```diff
@@ -170,21 +171,31 @@ private JvmObjectReference CreateBroadcast(SparkContext sc, T value)
             bool encryptionEnabled = bool.Parse(
                 sc.GetConf().Get("spark.io.encryption.enabled", "false"));
 
+            var _pythonBroadcast = (JvmObjectReference)javaSparkContext.Jvm.CallStaticJavaMethod(
+                "org.apache.spark.api.python.PythonRDD",
+                "setupBroadcast",
+                _path);
+
             if (encryptionEnabled)
             {
-                throw new NotImplementedException("Broadcast encryption is not supported yet.");
+                var pair = (JvmObjectReference[])_pythonBroadcast.Invoke("setupEncryptionServer");
+
+                using ISocketWrapper socket = SocketFactory.CreateSocket();
+                socket.Connect(
+                    IPAddress.Loopback,
+                    (int)pair[0].Invoke("intValue"),
+                    (string)pair[1].Invoke("toString"));
+                var chunked = new ChunkedStream(socket.OutputStream, 8192);
```

Comment:
I'm not convinced we need this `ChunkedStream`.  I know the purpose of it is to maintain a buffer and as long as we keep writing data to the `ChunkedStream` object, it will make sure that the buffer is filled before writing to the stream.  And when we close it, it will flush the contents to the stream.  This is useful if you are planning to write multiple `value`s while reusing the same `ChunkedStream` object.  However, here we aren't reusing the `ChunkedStream` object only one `value` is written and then the `ChunkedStream` is immediately closed.  I don't see the value in maintaining this `buffer`.  The current implementation is effectively serializing `value` and copies its byte array 2x in memory before it writes to the `OutputStream`.

Reply:
I think the purpose of this in `python` is that it is writing the broadcast object to the stream in chunks <ins>**as**</ins> the object is being serialized (pickled).

It seems with the current state of your `ChunkedStream`, it would be equivalent to `python` first serializing(pickling) the object and <ins>**after**</ins> it's done, it would then write it to the stream in chunks.

Let me know if that's not the case here.

Reply:
The difference in writing straight to memory and what is happening through `ChunkedStream` is the fact that each chunk is being preceded by the size of chunks (8192 in this case) in the memory. Also in our scenario it is not coming in use because we are broadcasting a small value, but if the user is broadcasting a large value, the stream would have to be divided into chunks and written in a format that Scala can dechunk it.
This is essentially what happens in Python as well, Python first pickles the broadcast value, and then writes in fixed chunks, pre-pending it with the chunk size in the Stream, which is what Scala expects and needs in order to de-chunk it.

Reply:
I understand you have to prepend by the data by the size of the bytes, that's not the main issue I'm trying to point out.  Python is writing to the `ChunkedStream` as the obj is in the process of being pickled.  It's writing pickle protocol data, pers ids, opcodes, packed bytes, etc, and each of these are being written to the `ChunkedStream` as the pickling is occurring.

Even if our broadcast value was large, it would not be taking advantage of the current implementation of `ChunkedStream`, since you are have serialized the whole value into a byte array before you start processing it. `            byte[] bytes = ConvertToByteArray(value);`

For reference:
[pickle.py 2.7](https://github.com/python/cpython/blob/2.7/Lib/pickle.py)
[pickle.py 3.8](https://github.com/python/cpython/blob/3.8/Lib/pickle.py)

Reply:
hmm I see... in that case I should find a way to serialize the broadcast value while writing it to the `ChunkedStream`?

Reply:
Removed `ChunkedStream` and added a function to do the copying logic, thanks!